src/pybind/mgr/dashboard/models_proposal.md

```diff
@@ -0,0 +1,608 @@
+  security_label=True,
+  protocols=[3, 4],
+  transports=["TCP", "UDP"])
+
+export.fsal = FsalModel(
+  name="CEPH",
+  user_id="fs_a",
+  fs_name="a",
+  sec_label_xattr="security.selinux")
+
+export.clients.append(ClientModel(
+  addresses=["192.168.100.0/24"],
+  access_type="RO",
+  squash="no_root_squash"
+))
+
+export.clients.append(ClientModel(
+  addresses=["192.168.1.103", "192.168.1.104"],
+  access_type="RW",
+  squash="all_squash"
+))
+```
+
+Model validations are triggered automatically by the controller infrastructure before calling the
+endpoint method. This way when the endpoint method is called the model was already validated, which
+can reduce a lot of validation code in the controller.
+Any validation error found in a model will originate an HTTP 400 (Bad Request) response.
+
+
+## Models for path, query, and header data
+
+### Path parameters
+
+We can also use model classes to represent path parameters. Path parameters correspond to data
+passed in the URL of the request and is usually used as the key of the resources.
+
+Currently, path parameters are encoded by our controllers as python method parameters following
+some rules as specified in the `HACKING.rst` file. One of the main limitations of current approach
+is that the ordering or position of the method parameters in the endpoint methods matters to
+distinguish path parameters from body or query parameters.
+
+By using a model class extended by `PathModel`, this limitation disappears because the path
+parameters become explicit in the endpoint method declaration, and the position of the parameter
+does not matter anymore. Moreover, we also gain the automatic validation of the path parameter
+value, as well as, the correct encoding of the path value string into the respective type.
+
+Let's continue with the nfs-ganesha example to show how to use path models.
+
+Currently the nfs-ganesha controller code is the following:
+
+```python
+@ApiController('/nfs-ganesha/export', Scope.NFS_GANESHA)
+class NFSGaneshaExports(RESTController):
+  RESOURCE_ID = "cluster_id/export_id"
+  # ...
+
+  @NfsTask('edit', {'cluster_id': '{cluster_id}', 'export_id': '{export_id}', 2.0)
+  def set(self, cluster_id, export_id, path, daemons, pseudo, tag, access_type,
+          squash, security_label, protocols, transports, fsal, clients,
+          reload_daemons=True):
+    export_id = int(export_id)
+    ganesha_conf = GaneshaConf.instance(cluster_id)
+    # ...
+    return {
+      "export_id": 1,
+      #...
+    }
+```
+
+As you can see from above the two first parameters of the `set` method correspond to the path
+parameters declared in the class attribute `RESOURCE_ID`. These parameters must appear before the
+parameters used for the body data in the endpoint method declaration.
+
+With the new path models we can rewrite the above example as the following:
+
+```python
+class ExportKeyModel(PathModel):
+  cluster_id = Attr.String(
+    description="Cluster identifier")
```

Comment:
Once again, this is just an example and not a methodology on how represent primary keys of data. 

Regarding the DataMapper vs ActiveRecord topic, the ActiveRecord was what we had in openATTIC and it was really painful in several aspects, including development of new models and performace. The main problem is that the backend is not consuming data from a single kind of source (like a relational database, or a key-value store), it's getting data from completely different kinds of sources. Trying to create a one-size-fits-all solution does not work for this case because what works good for one kind of source does not work quite well with other kinds of sources.

This is why I state that the "models"  in this proposal are only to be used to specify the structure of data of the endpoints and not in the services. I believe the best solution is what we currently have that is custom code for each service.

Reply:
ORMs have cultivated a bad reputation of giving birth to bloated and underperforming solutions, but that claim doesn't differ too much from the ones that are blamed on: OOP/C++, relational databases, NoSQL databases, interpreted languages, distributed systems, etc. And here we're using a bunch of them.

That said, I've partaken and suffered in the past from the pains of badly designed/implemented ORM code. However, IMHO it had more to do with the lack of proper documentation, learning/knowledge sharing, and poor debuggability/logging/profiling groundworks, than to the ORM concept per se. I don't know your pains with OpenATTIC and Django, but I'd really like to know, as I find that kind of pain-sharing very enriching (and therapeutic).

I'm really sorry to disagree on the last part, but I don't think our current services implementation is the best we can have and do. With our current approach, as we lack a mapping between inbound and outbound data (something Models essentially mean to provide), we cannot (or hardly can) do essential things to ensure a performant API:
- Lazy loading/fetching: only the data requested on the controller side is fetched from the back-end. A Model basically provides the best interface for that.
  - Without this it's really hard to implement pagination and sorting.
- Fine-grained caching: per-piece-of-information caching with different policies and expiration times.